### PR TITLE
Serialise and deserialise `SkinInfo.InstantiationInfo` to allow for more correct imports

### LINKED ISSUE #15896: Modified default skin exports don't import with the correct `InstantiationInfo`
The type is not serialised anywhere on export, so it's not feasible to correctly import such cases back.

Currently all imported skins get the default type as per:

https://github.com/ppy/osu/blob/f6a3709060b3ba4d81ce4f9d01df52aba7fa18dd/osu.Game/Skinning/SkinInfo.cs#L40-L45

### Discussed in https://github.com/ppy/osu/discussions/15895

<div type='discussions-op-text'>

<sup>Originally posted by **puyoxyz** December  2, 2021</sup>
I tried importing an OSK file with *just the JSON file with the HUD layout* and it works, but it's using the classic default skin. Is there a way to import a JSON HUD layout into the osu!lazer skin, or fall back to osu!lazer skin elements instead of the classic skin elements?</div>

### PR DESCRIPTION
**Pull request by peppy**
Until now, skins were always imported using the `LegacySkin` instantiation type. For cases where a user has edited the lazer or classic default (via the new skin editor), which would result in incorrect fallback paths after exporting and importing the edited skin.

- Closes #15896
- [x] Depends on #15859.

### CODE REVIEW

File: osu.Game/Skinning/SkinModelManager.cs [hunk 1]
```diff
@@ -49,8 +51,36 @@ public SkinModelManager(Storage storage, RealmContextFactory contextFactory, Gam
...
+                {
+                    LogForModel(model, $"Error during {skin_info_file} parsing, falling back to default", e);
+
+                    // Not sure if we should still run the import in the case of failure here, but let's do so for now.
+                    model.InstantiationInfo = string.Empty;
+                }
+            }
+
+            // Always rewrite instantiation info (even after parsing in from the skin json) for sanity.
+            model.InstantiationInfo = createInstance(model).GetType().GetInvariantInstantiationInfo();
```
Comment (bdach) [inline osu.Game/Skinning/SkinModelManager.cs:83]: Is this just to fix up the potential `string.Empty` case if an exception is thrown above, or does this serve some secondary purpose?

File: osu.Game/Skinning/SkinModelManager.cs [hunk 2]
```diff
@@ -49,8 +51,36 @@ public SkinModelManager(Storage storage, RealmContextFactory contextFactory, Gam
...
+            var skinInfoFile = model.Files.SingleOrDefault(f => f.Filename == skin_info_file);
+
+            if (skinInfoFile != null)
+            {
+                try
+                {
+                    using (var existingStream = Files.Storage.GetStream(skinInfoFile.File.GetStoragePath()))
+                    using (var reader = new StreamReader(existingStream))
+                    {
+                        var deserialisedSkinInfo = JsonConvert.DeserializeObject<SkinInfo>(reader.ReadToEnd());
```
Comment (bdach) [inline osu.Game/Skinning/SkinModelManager.cs:63]: While this works, I'd say that maybe it's dumping a bit too much info. Here's an example to show what I mean:

```json
{
  "Name": "osu! (triangles) (modified)",
  "Creator": "Unknown",
  "Hash": "474e3ae1925613810d7c49b1c8bed2946ef21fd6d6f999f66a9939b219f44388", // probably shouldn't be here
  "Protected": false, // probably shouldn't be here
  "InstantiationInfo": "osu.Game.Skinning.DefaultSkin, osu.Game",
  "Files": [ // this entire property is a little redundant
    {
      "File": {
        "Hash": "0108d0eb129aea08415719722873bc29805ada109479a33b78fe8ff1e82a4008" // probably shouldn't be here
      },
      "Filename": "skin.ini",
      "Parent": null // inherited from EmbeddedObject, probably shouldn't be here
    },
    {
      "File": {
        "Hash": "c7bc3465f74ffdfc545f999b5cb4f64a545da90cdccccdceca963fb8c2c8e74a"
      },
      "Filename": "skininfo.json",
      "Parent": null
    },
    {
      "File": {
        "Hash": "31cd6aa52bd39ec58d874bba33857781bf6d392a5a0b6e7feba0e7976f94e368"
      },
      "Filename": "MainHUDComponents.json",
      "Parent": null
    }
  ],
  "DeletePending": false // probably shouldn't be here
}
```

File: osu.Game/Skinning/SkinModelManager.cs (hunk 1 above)
Comment (peppy) [reply osu.Game/Skinning/SkinModelManager.cs:83]: Basically yep.

File: osu.Game/Skinning/SkinModelManager.cs (hunk 2 above)
Comment (peppy) [reply osu.Game/Skinning/SkinModelManager.cs:63]: Yeah, I have a TODO which is to go through each serialisable model and make them explicit include only for json, but it does make sense to fix this one here while we're at it I guess.

Comment (peppy) [reply osu.Game/Skinning/SkinModelManager.cs:63]: I've made this opt-in now, looks like this:

```json
{
  "Name": "osu! (triangles) (modified)",
  "Creator": "Unknown",
  "InstantiationInfo": "osu.Game.Skinning.DefaultSkin, osu.Game"
}
```

**Review by bdach — APPROVED** [review]

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**